### PR TITLE
Reapply D14078519

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17596 Reapply D14078519**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14270288/)

Was reverted before, now fixed version.

Differential Revision: [D14270288](https://our.internmc.facebook.com/intern/diff/D14270288/)

### CODE REVIEW

File: caffe2/operators/reduction_ops.h [hunk 1]
```diff
@@ -14,12 +14,16 @@ class SumElementsOp : public Operator<Context> {
...
+  explicit SumElementsOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
         average_(this->template GetSingleArgument<bool>("average", false)) {}
   explicit SumElementsOp(const OperatorDef& operator_def, Workspace* ws, bool average)
       : Operator<Context>(operator_def, ws), average_(average) {}
+  explicit SumElementsOp(const c10::FunctionSchema& schema, std::vector<c10::IValue> inputs, std::vector<c10::IValue*> outputs)
+      : Operator<Context>(schema, std::move(inputs), std::move(outputs)),
+        average_(this->template GetSingleArgument<bool>("average", false)) {}
+  explicit SumElementsOp(const c10::FunctionSchema& schema, std::vector<c10::IValue> inputs, std::vector<c10::IValue*> outputs, bool average)
+      : Operator<Context>(schema, std::move(inputs), std::move(outputs)), average_(average) {}
```
Comment (ezyang) [inline caffe2/operators/reduction_ops.h]: What's going on here?

File: caffe2/operators/reduction_ops.h [hunk 2]
```diff
@@ -76,12 +80,16 @@ class SumElementsGradientOp : public Operator<Context> {
...
+  explicit SumElementsGradientOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
         average_(this->template GetSingleArgument<bool>("average", false)) {}
   explicit SumElementsGradientOp(const OperatorDef& operator_def, Workspace* ws, bool average)
       : Operator<Context>(operator_def, ws), average_(average) {}
+  explicit SumElementsGradientOp(const c10::FunctionSchema& schema, std::vector<c10::IValue> inputs, std::vector<c10::IValue*> outputs)
+      : Operator<Context>(schema, std::move(inputs), std::move(outputs)),
+        average_(this->template GetSingleArgument<bool>("average", false)) {}
+  explicit SumElementsGradientOp(const c10::FunctionSchema& schema, std::vector<c10::IValue> inputs, std::vector<c10::IValue*> outputs, bool average)
+      : Operator<Context>(schema, std::move(inputs), std::move(outputs)), average_(average) {}
```
Comment (ezyang) [inline caffe2/operators/reduction_ops.h]: Ditto

**Review by ezyang — APPROVED** [review]

File: caffe2/operators/reduction_ops.h (hunk 1 above)
Comment (smessmer) [reply caffe2/operators/reduction_ops.h]: SumElementsOp() is both, an operator and a base class for other operators. When used as an operator, it is constructed with `(operator_def, workspace)`, when used as a base class it is constructed with `(operator_def, workspace, average)`. So unfortunately, one constructor that just passes through all arguments to the `Operator` constructor doesn't work here and I need to spell out all variants.

File: caffe2/operators/reduction_ops.h (hunk 2 above)
Comment (smessmer) [reply caffe2/operators/reduction_ops.h]: see comment above

**Pull request closed without merging**